# Model Building

Models to be used
- 1) Logistic Regression 
- 2) SVM 
- 3) KNN
- 4) Decision Tree
- 5) Random Forest

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# data for model building
data_train = pd.read_csv('/Users/amit/Documents/Project Hamoye/Model Building/train_data.csv')
data_train.head()  

Visit   MR Delay  Age  EDUC  eTIV   nWBV    ASF  SES_imputed  MMSE_imputed  \
0      1   0.000000   70    11  1445  0.722  1.214          4.0          22.0   
1      4  31.144823   68    16  1714  0.682  1.024          1.0           7.0   
2      3  39.887341   82    18  1442  0.644  1.217          1.0          29.0   
3      2  21.377558   94    23  1474  0.696  1.190          1.0          29.0   
4      2  26.457513   72    16  1796  0.732  0.977          4.0          28.0   

   missing_indicator_SES  Group_Demented  Group_Nondemented  M/F_F  CDR  
0                      0               1                  0      1  1.0  
1                      0               1                  0      0  1.0  
2                      0               0                  1      1  0.0  
3                      0               0                  1      1  0.0  
4                      0               1                  0      0  0.5

In [3]:
data_train.shape

(298, 14)

In [4]:
# data for testing the model
data_test = pd.read_csv('/Users/amit/Documents/Project Hamoye/Model Building/test_data.csv')
data_test.head()

Visit   MR Delay  Age  EDUC  eTIV   nWBV    ASF  SES_imputed  MMSE_imputed  \
0      3  37.881394   73    12  1478  0.696  1.188          4.0          16.0   
1      1   0.000000   66    12  1446  0.780  1.214          4.0          30.0   
2      4  42.071368   73    13  1360  0.773  1.291          4.0          30.0   
3      1   0.000000   71    14  1461  0.756  1.202          2.0          28.0   
4      5  48.989795   86    12  1813  0.761  0.968          3.0          27.0   

   missing_indicator_SES  Group_Demented  Group_Nondemented  M/F_F  CDR  
0                      0               1                  0      1  1.0  
1                      0               1                  0      0  0.5  
2                      0               0                  1      1  0.0  
3                      0               0                  1      1  0.0  
4                      0               0                  1      0  0.0

In [5]:
data_test.shape

(75, 14)

In [6]:
X_train , X_test = data_train.drop('CDR' , axis=1) , data_test.drop('CDR' , axis=1)
y_train , y_test = data_train['CDR'].astype(int) , data_test['CDR'].astype(int) 
# test converted to int because the classification doesnot accept float dtype 

Divided the train and test data to use for the models 

In [7]:
X_train.head()

Visit   MR Delay  Age  EDUC  eTIV   nWBV    ASF  SES_imputed  MMSE_imputed  \
0      1   0.000000   70    11  1445  0.722  1.214          4.0          22.0   
1      4  31.144823   68    16  1714  0.682  1.024          1.0           7.0   
2      3  39.887341   82    18  1442  0.644  1.217          1.0          29.0   
3      2  21.377558   94    23  1474  0.696  1.190          1.0          29.0   
4      2  26.457513   72    16  1796  0.732  0.977          4.0          28.0   

   missing_indicator_SES  Group_Demented  Group_Nondemented  M/F_F  
0                      0               1                  0      1  
1                      0               1                  0      0  
2                      0               0                  1      1  
3                      0               0                  1      1  
4                      0               1                  0      0

In [8]:
y_train.head()

0    1
1    1
2    0
3    0
4    0
Name: CDR, dtype: int64

In [9]:
y_train.value_counts()

0    261
1     34
2      3
Name: CDR, dtype: int64

## 1) Logistic Regression

In [10]:
from sklearn.linear_model import LogisticRegression
model_logistic = LogisticRegression()

__HypterParameter Optimisation__

In [11]:
grid_params = {
    'penalty':['l1', 'l2', 'elasticnet', 'none'] , 
    'dual':[True ,False] , 
    'solver':['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'] , 
    'max_iter':[i for i in np.arange(10 , 100 , 90)]
}


from sklearn.model_selection import RandomizedSearchCV
randomised_search_cv = RandomizedSearchCV(estimator=model_logistic , param_distributions=grid_params , 
                                         n_iter=100 , scoring='accuracy' , n_jobs=-1 , cv=10 , 
                                         verbose=2 , random_state=2)
randomised_search_cv.fit(X_train , y_train)

Fitting 10 folds for each of 40 candidates, totalling 400 fits


/opt/anaconda3/envs/tf/lib/python3.7/site-packages/sklearn/model_selection/_search.py:296: UserWarning: The total space of parameters 40 is smaller than n_iter=100. Running 40 iterations. For exhaustive searches, use GridSearchCV.
  UserWarning,
/opt/anaconda3/envs/tf/lib/python3.7/site-packages/sklearn/model_selection/_split.py:680: UserWarning: The least populated class in y has only 3 members, which is less than n_splits=10.
  UserWarning,
/opt/anaconda3/envs/tf/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: 
280 fits failed out of a total of 400.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
20 fits failed with the following error:
Traceback (most recent call last):
  File "/opt/anaconda3/

RandomizedSearchCV(cv=10, estimator=LogisticRegression(), n_iter=100, n_jobs=-1,
                   param_distributions={'dual': [True, False], 'max_iter': [10],
                                        'penalty': ['l1', 'l2', 'elasticnet',
                                                    'none'],
                                        'solver': ['newton-cg', 'lbfgs',
                                                   'liblinear', 'sag',
                                                   'saga']},
                   random_state=2, scoring='accuracy', verbose=2)

In [12]:
randomised_search_cv.best_params_     # parameters selected after Randomised Search

{'solver': 'liblinear', 'penalty': 'l1', 'max_iter': 10, 'dual': False}

In [13]:
randomised_search_cv.best_score_    # best acccuracy from our model after the cross validation on train data

0.8960919540229886

In [14]:
# Model Logistic Regression with the best parameters selected 
model_logistic = LogisticRegression(solver='liblinear',penalty='l1' , max_iter= 10, dual= False , random_state=2)
model_logistic.fit(X_train , y_train)

/opt/anaconda3/envs/tf/lib/python3.7/site-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,


LogisticRegression(max_iter=10, penalty='l1', random_state=2,
                   solver='liblinear')

In [15]:
predict_train = model_logistic.predict(X_train)
predict_test = model_logistic.predict(X_test)

In [16]:
from sklearn.metrics import accuracy_score
print('The accuracy of our model on train and test data')
print('Train : ' , accuracy_score(predict_train , y_train))
print('Test : ' , accuracy_score(predict_test , y_test))

The accuracy of our model on train and test data
Train :  0.9093959731543624
Test :  0.9466666666666667


In [17]:
from sklearn.metrics import confusion_matrix
confusion_matrix(predict_test , y_test)

array([[68,  4],
       [ 0,  3]])

So our model is woking great with the train accuracy of 90% and test accuracy of 94%

## 2) SVM

## 3) KNN

## 4) Decision Tree

## 5) Random Forest